### ISSUE TITLE
Scene file format? 

### ISSUE THREAD

**Issue by JFrap** [JFrap/TeazelEngine#28]
I presume we'll go binary, how should we construct the format?

**Comment by unpause** [comment]
as the structure for the format i was thinking about maybe this: 

`te_scene:
    version=ENGINE_VERSION
    name=SCENE_NAME
    id=SCENE_ID
    settings:
        PER_SCENE_SETTINGS_HERE
    resources:
        ext_res="res://res1.res" id=0 // .res file is a meta file that refers
        to some other file and has information about it in it.
        ext_res="res://res2.res" id=1
        sub_res=SUB_RES_DATA
        ...
        
    scene_tree:
        object:
            name=OBJECT_NAME
            id=OBJECT_ID // This is a guid which is unique per object and only exists once per project
            template_id=TEMPLATE_ID
            components:
                component:
                    name=COMPONENT_NAME
                    type=COMPONENT_TYPE
                    enabled=COMPONENT_ENABLED
                    component_var_1=COMPONENT_VAR
                    component_var_2=COMPONENT_VAR
                    component_var_3=COMPONENT_VAR
                    component_var_4=res(id=0)
                    ...
                ...
        ...
    `

this will be saved as a text file in whatever data format we want and will be compressed at editor project compile time into a packed file which gets then loaded when the game load the scene from the resource path and is then deserialized into a scene.

**Comment by unpause** [comment]
Specification about Parent-Children relationships and for Object Templates will follow

**Comment by JFrap** [comment]
Looks good.

**Comment by JFrap** [comment]
But, this is very per-scene, should we have a "Project" class that collects all the scenes, and indexes them in a file.

**Comment by JFrap** [comment]
And what should the file extension be? Ill suggest .ts (Teazel scene)

**Comment by unpause** [comment]
we will have a project file will only be for the editor later. the scene compiler in the editor will compile the files move them to the scene folder and rename them to `scene_id.tzs`. 

for file extensions i would name scenes `.tzs` template `.tzt` and so on

**Comment by unpause** [comment]
What idea do you have about the scene file structure?

**Comment by unpause** [comment]
What idea do you have about the scene file structure?

**Comment by unpause** [comment]
Updated for Parent-Children relationship.

**Comment by JFrap** [comment]
@unpause Something rather similar to the scene, but....

    version=ENGINE_VERSION
    name=SCENE_NAME
    id=SCENE_ID
    settings:
        PER_SCENE_SETTINGS_HERE
    resources:
        ext_res="res://res1.res" id=0 // .res file is a meta file that refers
        to some other file and has information about it in it.
        ext_res="res://res2.res" id=1
        sub_res=SUB_RES_DATA
        ...

        templates:
            object:
                name=TEMPLATE_NAME
                id=TEMPLATE_ID // This is a guid which is unique per object and only exists once per project
                parent=PARENT_ID // Should templates have parents? 
                template_id=TEMPLATE_ID
                components:
                    component:
                        name=COMPONENT_NAME
                        type=COMPONENT_TYPE
                        enabled=COMPONENT_ENABLED
                        component_var_1=COMPONENT_VAR
                        component_var_2=COMPONENT_VAR
                        component_var_3=COMPONENT_VAR
                        component_var_4=res(id=0)

                        ...
                ....

    scenes:
        scene:
            name=TEMPLATE_NAME
            id=SCENE_ID
            path=SCENE_PATH

**Comment by unpause** [comment]
i dont think templates should belong to a scene

**Comment by unpause** [comment]
Specification for meta files and editor project files:

```
// .tzm meta file
[file]
name=FILE_NAME
type=FILE_TYPE
// maybe more info here

// .tzp project file
[project]
name=PROJECT_NAME
engine_version=ENGINE_VERSION
files:
	LIST_OF_TZM_FILES
other_project_settings:
... // other project settings follow here
```

.tzm is a meta file which gets created for every normal asset file and holds some information for the editor about it. i think about maybe integrating the meta files into the project file.

.tzp is a project file which gets created by the editor. besides of information about the project it also holds information about the editor version with which the project was created so we can patch files if some breaking changes happened.
The project file also holds the relative path to every meta file so the editor can load all asset files easily without searching directories for files.
Then there will be a settings for project unique settings.

**Comment by unpause** [comment]
When the editor opens the project a project.lock file will be created which holds a timestamp it has the purpose to hinder other editors or people to open the project at the same moment and modify files of the priject and eventually damaging them.

**Comment by unpause** [comment]
btw we should do a discussion about how we want to de/serialize the components and the scene because i have no idea on how we want to do that. maybe @xforce has an idea what solutions exist for c++

**Comment by unpause** [comment]
We also need to think about the thing with the C++ Plugins how we want to register the Components and de/serialize them. I dont have that much experience with working with dymanic linking and stuff.

**Comment by JFrap** [comment]
"Dynamic linking" linking dll's at runtime?

**Comment by unpause** [comment]
yes thats basically dynamic linking. but thats not the problem. the problem is how we will handle de/serialization of Components and so on

**Comment by xforce** [comment]
The best serailization libraries I know are [cereal](http://uscilab.github.io/cereal/) and [Boost.Serailziation](http://www.boost.org/doc/libs/1_61_0/libs/serialization/doc/index.html) I have used both before and they both work very well.

Recently we also have been using [Protobuf](https://developers.google.com/protocol-buffers/) which uses proto files to 'describe' the object, Protobuf is a bit more work to use it though.

**Comment by unpause** [comment]
Well, we would need something that the user can either easily write himself or something that can get generated from the code files that the user writes.

**Comment by unpause** [comment]
We also should take into consideration that the scenes will probably be saved and edit by an editor so we also need some way to get what fields the components have in some way but thats probably nothing todo with file formats.

**Comment by unpause** [comment]
Example in XML:
```xml
<?xml version="1.0"?>

<TE_SCENE>

	<version>0.1.0</version>
	<name>SceneName</name>
	<id>1</id>
	
	<scene_tree>
	
		<object>
			<id></id>
			<components>
				<name>TransformComponent</name>
				<enabled>true</enabled>
				<var="position">
					<x>1</x>
					<y>1</y>
					<z>1</z>
				</var>
				<var="rotation">
					<x>0</x>
					<y>0</y>
					<z>0</z>
				</var>
			</components>
		</object>
		
	</scene_tree>
	
</TE_SCENE>
```
